pkg/lib/ldap/client.go

```diff
@@ -0,0 +1,109 @@
+package ldap
+
+import (
+	"bytes"
+	"errors"
+	"html/template"
+	"net/url"
+	"strings"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/go-ldap/ldap/v3"
+)
+
+type UserProfile *ldap.Entry
+
+type LDAPClient struct {
+	Config config.LDAPServerConfig
+	conn   *ldap.Conn
+}
+
+func (c *LDAPClient) Connect() error {
+	conn, err := ldap.DialURL(c.Config.URL)
+	if err != nil {
+		return err
+	}
+
+	url, err := url.Parse(c.Config.URL)
+	if err != nil {
+		return err
+	}
+	if url.Scheme == "ldap" {
+		_ = conn.StartTLS(nil)
+	}
+
+	c.conn = conn
+	return nil
+}
+
+func (c *LDAPClient) Close() {
+	if c.conn != nil {
+		c.conn.Close()
+	}
+}
+
+func (c *LDAPClient) Bind(username string, password string) error {
+	if c.conn == nil {
+		return errors.New("ldap: not connected")
+	}
+	return c.conn.Bind(username, password)
+}
+
+func parseSearchFilter(filter string, username string) (string, error) {
+
+	trimedFilter := strings.TrimSpace(filter)
+
+	tmpl, err := template.New("search_filter").Parse(strings.Trim(trimedFilter, "\n"))
+	if err != nil {
+		return "", err
+	}
+	var buf bytes.Buffer
+	// make a string array of phone, username, email
+	err = tmpl.Execute(&buf, map[string]string{"Username": username})
+
+	if err != nil {
+		return "", err
+	}
+	// check if the filter is correct
+	result := buf.String()
+	result = strings.TrimSpace(result)
+
+	_, err = ldap.CompileFilter(result)
+	if err != nil {
+		return "", err
+	}
+
+	return result, nil
+}
+
+func (c *LDAPClient) AuthenticateUser(username string, password string) (*ldap.Entry, error) {
+	searchFilter, err := parseSearchFilter(c.Config.SearchFilterTemplate, username)
+	if err != nil {
+		return nil, err
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		c.Config.BaseDN,
+		ldap.ScopeWholeSubtree, ldap.DerefAlways, 2, 0, false,
```

Comment:
I think we need to add comment to remind us why the forth param is 2

Time limit should not be infinite.

and why false

Reply:
SearchRequest.typesOnly: An indicator as to whether Search results are to contain both
   attribute descriptions and values, or just attribute descriptions.
   Setting this field to TRUE causes only attribute descriptions (and
   not values) to be returned.  Setting this field to FALSE causes both
   attribute descriptions and values to be returned.
Hence we set to false for getting both attribute descriptions and values.